src/pydash/collections.py

```diff
@@ -1059,6 +2258,24 @@ def some(collection, predicate=None):
     return any(collection)
 
 
+@t.overload
+def sort_by(
+    collection: t.Mapping[t.Any, T2],
+    iteratee: t.Union[t.Callable[[T2], t.Any], IterateeObjT, None] = None,
+    reverse: bool = False,
+) -> t.List[T2]:
+    ...
+
+
+@t.overload
+def sort_by(
+    collection: t.Iterable[T],
+    iteratee: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
+    reverse: bool = False,
+) -> t.List[T]:
+    ...
+
+
```

Comment:
All these overloads are contributing quite a bit to total line length. Are they all necessary? Wondering if being less strict about objects containing all of the same type would eliminate the need for some of these overloads?

Reply:
The main idea of most of these overloads is to separate mappings from iterables. IMO this is needed if we want to type the iteratees properly. `Iterable` is not a `Mapping` and `Mapping` is an `Iterable` but it is not generic on the value of the mapping.

```python
@t.overload
def sort_by(
    collection: t.Mapping[t.Any, T2],
    iteratee: t.Union[t.Callable[[T2], t.Any], IterateeObjT, None] = None,
    reverse: bool = False,
) -> t.List[T2]:
    ...


@t.overload
def sort_by(
    collection: t.Iterable[T],
    iteratee: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
    reverse: bool = False,
) -> t.List[T]:
    ...
```
So the idea here is that without overload we have no way to say this function accepts any iterable and the iteratee's first argument (if it is a callable) will be the value of the mapping if we use a mapping as `collection`

This is because pydash passes the value of the mapping as first argument to the iteratee which is nice but it makes typing a bit trickier.

Reply:
I was only able to skim the overloads on my first pass through so I was mostly concerned with how many there were.

Looking at this particular example of `sort_by`, could that be simplified to:

```python
def sort_by(
    collection: t.Union[t.Iterable[T], t.Mapping[t.Any, T]],
    iteratee: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
    reverse: bool = False,
) -> t.List[T]:
    ...
```

If so, are there other functions where we could potentially use `t.Union` to reduce the number of overloads in cases where the other arguments are common between them?

Reply:
Unfortunately this doesn't really work because `Iterable[T]` and `Mapping[t.Any, T]` overlap.
`pyright` seems to pick the less specific, so in this case if we pass a dict, the `Iterable[T]` will be picked and `T` will be the keys' type instead of the values' type.
`mypy` rejects this type altogether, passing a dict, we get an incompatible type error, not exactly sure why.

When I first started typing this library I used a generic type alias for the iteratee type which reduced overload amount a lot. If we pick `drop_while` as a simple example. Current type is:
```python
@t.overload
def drop_while(array: t.List[T], predicate: t.Callable[[T, int, t.List[T]], t.Any]) -> t.List[T]:
    ...


@t.overload
def drop_while(array: t.List[T], predicate: t.Callable[[T, int], t.Any]) -> t.List[T]:
    ...


@t.overload
def drop_while(array: t.List[T], predicate: t.Callable[[T], t.Any]) -> t.List[T]:
    ...


@t.overload
def drop_while(array: t.List[T], predicate: None = None) -> t.List[T]:
    ...
```

And with my generic type alias idea I had:
```python
F = t.Union[
    t.Callable[[T, T2, t.List[T]], T3],
    t.Callable[[T, T2], T3],
    t.Callable[[T], T3],
    None,
]

def drop_while(array: t.List[T], predicate: F[T, int, t.Any] = None):
    ...
```

Which is really nice, you can reduce a lot of overload, and it works for list and dict, depending on the generic values you pass. Here `int` as second because it is a list, for mapping it would be:

```python
F = t.Union[
    t.Callable[[T, T2, t.Dict[T2, T]], T3],
    t.Callable[[T, T2], T3],
    t.Callable[[T], T3],
    None,
]

def some_func_accepting_dict(d: t.Mapping[T2, T], predicate: F[T, T2, t.Any] = None):
    ...
```

This is a bit different than the `sort_by` example above as you still need to have overloads for `Iterable` and `Mapping` but it combines all the different predicate/iteratee overloads in one single overload.

While this works perfectly fine with `pyright`, it doesn't work at all with `mypy`, it just doesn't get it somehow and everything is typed as `Any`. I opened an issue in `mypy` about this when I realized: https://github.com/python/mypy/issues/14498

Reply:
Does mypy make any guarantees around which type it uses from a Union when an arg matches more than one? Like use the first matched from the Union list or anything like that? So if the more specific match like Mapping was first in the Union would that one be used over any other types in the Union? Or is behavior undefined for this situation and can't be relied on?

Reply:
More than undefined behaviour, it doesn't even accept it if both types in the union overlaps it seems.
Small test:
```python
# test_type.py
import typing as t

T = t.TypeVar("T")

def func(x: t.Union[t.Mapping[t.Any, T], t.Iterable[T]]) -> t.List[T]:
    return []

d: t.Dict[int, str] = {5: "hello"}

x = func(d)
reveal_type(x)
```
While `pyright` accepts this but matches the wider type (`Iterable`), we get revealed type `x: List[int]`
`mypy` doesn't even accept it:
```bash
$ mypy test_type.py
test_type.py:14: error: Need type annotation for "x" (hint: "x: List[<type>] = ...")  [var-annotated]
test_type.py:14: error: Argument 1 to "func" has incompatible type "Dict[int, str]"; expected "Union[Mapping[Any, <nothing>], Iterable[<nothing>]]"  [arg-type]
test_type.py:15: note: Revealed type is "builtins.list[Any]"
Found 2 errors in 1 file (checked 1 source file)
```

What seems to work is to restrict the iterable type:
```python
def sort_by(
    collection: t.Union[t.List[T], t.Mapping[t.Any, T]],
    iteratee: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
    reverse: bool = False,
) -> t.List[T]:
    ...
```

Reply:
So potentially unions of List/Mapping or Sequence/Mapping might work? How much would that potentially reduce the number of overloads needed across functions? If it's significant then that might be something to compromise on. Otherwise, may just bite the bullet and have the overloads to get a little better on the typing.

Reply:
I don't think this would reduce much. This would only be possible in functions that don't accept the full iteratee as implemented in the combination of `iteriteratee` and `callit`. The main thing that requires all these overloads is the combinations of functions accepting `Iterable` or `Mapping` and the different number of arguments the `predicate`/`iteratee` can be called with, as I showed in my last comment.
`sort_by` is an example that don't accept the full arguments iteratee, so it has only 2 overloads and using List/Mapping or Sequence/Mapping would reduce to only 1 overload (eg. no overload needed).
But most functions are typed like this `reject` for example:

```python
@t.overload
def reject(
    collection: t.Mapping[T, T2],
    predicate: t.Union[t.Callable[[T2, T, t.Dict[T, T2]], t.Any], IterateeObjT, None] = None,
) -> t.List[T2]:
    ...


@t.overload
def reject(
    collection: t.Mapping[T, T2],
    predicate: t.Union[t.Callable[[T2, T], t.Any], IterateeObjT, None] = None,
) -> t.List[T2]:
    ...


@t.overload
def reject(
    collection: t.Mapping[t.Any, T2],
    predicate: t.Union[t.Callable[[T2], t.Any], IterateeObjT, None] = None,
) -> t.List[T2]:
    ...


@t.overload
def reject(
    collection: t.Iterable[T],
    predicate: t.Union[t.Callable[[T, int, t.List[T]], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...


@t.overload
def reject(
    collection: t.Iterable[T],
    predicate: t.Union[t.Callable[[T, int], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...


@t.overload
def reject(
    collection: t.Iterable[T],
    predicate: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...
```

And since in most of the overloads the signature of the `predicate` changes depending on the `collection` type, I think we can only remove 1 overload:
```python
@t.overload
def reject(
    collection: t.Mapping[T, T2],
    predicate: t.Union[t.Callable[[T2, T, t.Dict[T, T2]], t.Any], IterateeObjT, None] = None,
) -> t.List[T2]:
    ...


@t.overload
def reject(
    collection: t.Mapping[T, T2],
    predicate: t.Union[t.Callable[[T2, T], t.Any], IterateeObjT, None] = None,
) -> t.List[T2]:
    ...


@t.overload
def reject(
    collection: t.Iterable[T],
    predicate: t.Union[t.Callable[[T, int, t.List[T]], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...


@t.overload
def reject(
    collection: t.Iterable[T],
    predicate: t.Union[t.Callable[[T, int], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...


@t.overload
def reject(
    collection: t.Union[t.Sequence[T], t.Mapping[t.Any, T]],
    predicate: t.Union[t.Callable[[T], t.Any], IterateeObjT, None] = None,
) -> t.List[T]:
    ...
```

Reply:
Alright, so sounds like we need the overloads as they are. The only other thing I can think of to help reduce the overall line count would be to collapse the overload functions defs to a single line and have the formatter ignore those lines. But I'm not sure much that would hurt readability of the types since some of them might be very difficult to grok if the function def and its typed arguments are on a single line. But for now, I think leaving it as-is will be fine.